### ISSUE TITLE
Comments for 2020-09-04-fall-2020-plans

### ISSUE THREAD

**Issue by remexre** [remexre/remexre.github.io#9]

**Comment by atennapel** [comment]
I would be interesting in learning about your different representation for datatypes. I have implemented something like "Gentle art of levitation", but I am interested in other variations on it. (https://github.com/atennapel/coc-os/blob/master/lib/fin.p)

**Comment by atennapel** [comment]
Also, all of the links in the "Formally Verifying StahlOS" section do not work. (I get 404 Not Found)

**Comment by remexre** [comment]
Fixed the links. Great to hear others are trying the same thing! I hadn't actually seen the "Generic constructors and eliminators from descriptions" paper before; I'll give it a read. My representation is the same one as used in "The practical guide to levitation," it's basically (in Coq):

```coq
Inductive Desc (ix : Type) : Type :=
  | Ret : ix -> Desc ix
  | Arg : forall A, (A -> Desc ix) -> Desc ix
  | Indx : ix -> Desc ix -> Desc ix
  | HIndx : Type -> ix -> Desc ix -> Desc ix.

Fixpoint Syn (ix : Type) (desc : Desc ix) (rec : ix -> Type) (i : ix) : Type :=
  match desc with
  | Ret     j   => (i = j)
  | Arg   A   d => { a : A & Syn ix (d a) rec i }
  | Indx    j d =>       rec j  * Syn ix d rec i
  | HIndx A j d => (A -> rec j) * Syn ix d rec i
  end.

Axiom Data : forall ix, Desc ix -> ix -> Type.
Axiom Con : forall {ix} {desc : Desc ix} {i},
  Syn ix desc (Data ix desc) i -> Data ix desc i.
```

Looks like this is a subset of your `IDesc`? (Well, I guess this is also the representation of the first half of Gentle.) I extended the `ind` from the non-indexed part of gentle to support indices [in Coq](https://git.sr.ht/~remexre/stahl/tree/8db7bd221f7ea6d4ac17d8e639a9be5180427612/exp/levitation/Gentle.v), but it's been too annoying to actually use so far...

**Comment by remexre** [comment]
After giving "Generic constructors and eliminators from descriptions" a read-through, I think I've actually implemented something quite similar when I was trying to make it more ergonomic: https://git.sr.ht/~remexre/stahl/tree/ebc41462a4407fdc3481eaa5470cf9bdebc09bb1/exp/levitation/Levitation.v

This inlined `ind` with `elim` though, and I ended up nixing it for being too complicated to want to implement it as a compiler builtin. This go-around, I was planning to implement acceptable-for-humans-to-use eliminators using Idris-inspired elaborator; I might try just implementing this instead, since it'll let me put off implementing the elaborator for longer.

**Comment by atennapel** [comment]
Right, yeah I agree it's nicer to keep the primitives simpler, and with that paper the users can implement the nicer eliminators themselves.

In my descriptions I added:
- ArgE for erased arguments
- FArg for non-dependent arguments (just to keep descriptions as first-order as possible)
- I'd like to implement quantitative type theory as well, which will add ArgL and FArgL for linear arguments (maybe RecL and HRecL as well?)

Technically induction on descriptions is enough as a primitive (because you can implement All and all using just description induction), but since the type and elimination of the fixed point operator relies on All and all you need those as primitives as well.

One idea I had was to index the fixpoint operator with the definitions for All and all, then they do not have to be primitives.
